protons/driver.py

```diff
@@ -1338,6 +1197,221 @@ def _get_potential_energy(self, context, state_index, group_index=0):
+                 ncmc_steps_per_trial=0, ncmc_timestep=1.0 * units.femtoseconds,
+                 maintainChargeNeutrality=False, cationName='Na+', anionName='Cl-', implicit=False):
+        """
+        Initialize a Monte Carlo sams_sampler driver for simulation of protonation states and tautomers.
+
+        Parameters
+        ----------
+        system : simtk.openmm.System
+            System to be titrated, containing all possible protonation sites.
+        temperature : simtk.unit.Quantity compatible with kelvin
+            Temperature at which the system is to be simulated.
+        pH : float
+            The pH at which the system is to be simulated.
+        prmtop : simtk.openmm.app.Prmtop
+            Parsed AMBER 'prmtop' file (necessary to provide information on exclusions)
+        cpin_filename : string
+            AMBER 'cpin' file defining protonation charge states and energies of amino acids
+        integrator : simtk.openmm.integrator
+            The integrator used for dynamics
+        pressure : simtk.unit.Quantity compatible with atmospheres, optional, default=None
+            For explicit solvent simulations, the pressure.
+        nattempts_per_update : int, optional, default=None
+            Number of protonation state change attempts per update call;
+            if None, set automatically based on number of titratible groups (default: None)
+        simultaneous_proposal_probability : float, optional, default=0.1
+            Probability of simultaneously proposing two updates
+        debug : bool, optional, default=False
+            Turn debug information on/off.
+        ncmc_steps_per_trial : int, optional, default=0
+            Number of steps per NCMC switching trial, or 0 if instantaneous Monte Carlo is to be used.
+        ncmc_timestep : simtk.unit.Quantity with units compatible with femtoseconds
+            Timestep to use for NCMC switching
+        maintainChargeNeutrality : bool, optional, default=True
+            If True, waters will be converted to monovalent counterions and vice-versa.
+        cationName : str, optional, default='Na+'
+            Name of cation residue from which parameters are to be taken.
+        anionName : str, optional, default='Cl-'
+            Name of anion residue from which parameters are to be taken.
+        implicit: bool, optional, default=False
+            Flag for implicit simulation. Skips ion parameter lookup.
+
+        Todo
+        ----
+        * Allow constant-pH dynamics to be initialized in other ways than using the AMBER cpin file (e.g. from OpenMM app; automatically).
+        * Generalize simultaneous_proposal_probability to allow probability of single, double, triple, etc. proposals to be specified?
+
+        """
+
+        # Set defaults.
+        # probability of proposing two simultaneous protonation state changes
```

Comment:
Good suggestion. I will probably already run into this when I put together the ffxml reader. If I find that some of these sections are entirely reusable, I can extract functions and move them to the base class so that both Amber and Forcefield drivers can access it.